nsxt/data_source_nsxt_upgrade_postcheck.go

```diff
@@ -0,0 +1,159 @@
+							Type:        schema.TypeString,
+							Description: "Status of execution of upgrade post-checks",
+							Computed:    true,
+						},
+						"details": {
+							Type:        schema.TypeString,
+							Description: "Details about current execution of upgrade post-checks",
+							Computed:    true,
+						},
+						"failure_count": {
+							Type:        schema.TypeInt,
+							Description: "Total count of generated failures or warnings in last execution of upgrade post-checks",
+							Computed:    true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceNsxtUpgradePostCheckRead(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	aggregateInfoClient := upgrade_unit_groups.NewAggregateInfoClient(connector)
+	component := d.Get("type").(string)
+	timeout := d.Get("timeout").(int)
+	interval := d.Get("interval").(int)
+	delay := d.Get("delay").(int)
+
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{model.UpgradeChecksExecutionStatus_STATUS_IN_PROGRESS},
+		Target:  []string{model.UpgradeChecksExecutionStatus_STATUS_COMPLETED},
+		Refresh: func() (interface{}, string, error) {
+			listRes, err := aggregateInfoClient.List(&component, nil, nil, nil, nil, nil, nil, nil)
+			if err != nil {
+				return nil, "", fmt.Errorf("failed to retrieve %s upgrade post-check results: %v", component, err)
+			}
+			for _, res := range listRes.Results {
+				if *res.PostUpgradeStatus.Status == model.UpgradeChecksExecutionStatus_STATUS_IN_PROGRESS {
+					return res.PostUpgradeStatus, model.UpgradeChecksExecutionStatus_STATUS_IN_PROGRESS, nil
+				}
+			}
+			// Return UpgradeChecksExecutionStatus_STATUS_COMPLETED here doesn't mean all UpgradeUnitGroup post-checks are completed.
+			// Just to stop polling the states as none of UpgradeUnitGroup is in in_progress state.
+			return listRes, model.UpgradeChecksExecutionStatus_STATUS_COMPLETED, nil
+		},
+		Timeout:      time.Duration(timeout) * time.Second,
+		PollInterval: time.Duration(interval) * time.Second,
+		Delay:        time.Duration(delay) * time.Second,
+	}
+
+	_, err := stateConf.WaitForState()
+	if err != nil {
+		return err
+	}
+
+	listRes, err := aggregateInfoClient.List(&component, nil, nil, nil, nil, nil, nil, nil)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve %s upgrade post-check result: %v", component, err)
+	}
+	var failedGroup []map[string]interface{}
+	for _, res := range listRes.Results {
+		failureCounts := 0
+		if res.PostUpgradeStatus.FailureCount != nil {
+			failureCounts = int(*res.PostUpgradeStatus.FailureCount)
+		} else if res.PostUpgradeStatus.NodeWithIssuesCount != nil {
+			failureCounts = int(*res.PostUpgradeStatus.NodeWithIssuesCount)
+		}
+		if res.PostUpgradeStatus.Status != nil && *res.PostUpgradeStatus.Status != model.UpgradeChecksExecutionStatus_STATUS_COMPLETED || failureCounts > 0 {
+			elem := make(map[string]interface{}, 4)
+			elem["status"] = *res.PostUpgradeStatus.Status
```

Comment:
sdk would do the pointer dereferencing while also checking for nil, so we usually don't dereference pointers in assignment to schema